### ISSUE TITLE
Custom domain handles for Bluesky => AP

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1150]
We currently support custom domain handles for web => fediverse. We should do the same for Bluesky => fediverse! Much of the plumbing is already there.

Corresponds to #826, which is equivalent for custom domain handles on accounts bridged into Bluesky.

(Note that this is just to support them at all. _Switching_ from  `@[domain]@bsky.brid.gy` handle to `[domain]@[domain]` has extra complications, tracked in #1074.)

**Comment by snarfed** [comment]
From @elfprince13 in https://github.com/snarfed/bridgy-fed/issues/826#issuecomment-2218108078:

> I currently control:
> 
> * @elfprince13@mumak.app (personal Mastodon instance)
> * @elfprince13.mumak.app (ATproto/Bluesky account)
> 
> Even a single-user Mastodon instance does not run particularly well on a small AWS instance, and it requires more attention than I want to allocate for it to constantly switch apps for posting, so I have been thinking of sunsetting the Mastodon account in favor of the Bluesky account; however, I would like to:
> 
> * retain my existing Mastodon followers if at all possible.
> * continue using @elfprince13@mumak.app for my handle, rather than the clunkier @elfprince13.mumak.app@elfprince13.mumak.app which the current domain redirect support appears to require.

**Comment by snarfed** [comment]
@elfprince13 the other difficulty here is that we currently only support custom fediverse usernames for web sites, not Bluesky accounts. I'm reluctant to add user-level settings, since they'd require an authed UI that I don't currently have and am unlikely to build. Maybe we could do it by reading the desired username or @-@ handle from your Bluesky profile bio though.

**Comment by elfprince13** [comment]
If there’s an unauthed approach that optimizes around the specific (fairly
common) pattern of user.domain.tld -> ***@***.*** or otherwise checks
for some configuration file from domain.tld (or even a TXT record in the
DNS) I would slightly prefer that to putting it in the bio.
-- 
~Thomas

**Comment by snarfed** [comment]
Agreed. Bio is awkward and annoying, DNS or HTTP is inaccessible, authed UI is a burden on my end. I haven't heard a good easy (or at least not hard) answer yet. 🤷

**Comment by frncs-eu** [comment]
Hey!
Thanks for pointing me to this thread  💣 
I'm curious, why is DNS/HTTP inaccessible?


My 2 cents:

I don't think the bio approach is a good idea – ideally the user should verify the domain they're claiming, otherwise one could just say they're `@bugsbunny@warnerbros.com` in their bio and there's no way to verify that, plus there's gonna be duplicates problems (2 users claiming the same domain in their bio). Also yes, it's just awkward.

What @elfprince13 proposes about checking some configuration file from domain.tld sounds ok to me and similar to what Bluesky already does. I don't know if it is _technically_ impossible, the doc for the webside -> AP mentions a similar approach using the `.well-known` path but surely you know why it can't be used in this scenario and I don't 😛

**Comment by snarfed** [comment]
We may be confusing two different things here: custom _domain_ handles, eg `@mydomain.com@mydomain.com`, and custom _usernames_ in handles, eg `@alice@mydomain.com`. Let's keep this issue focused on the former; I'll create a new issue for the latter.

**Comment by snarfed** [comment]
^ done!

@frncs-eu:

> I don't think the bio approach is a good idea – ideally the user should verify the domain they're claiming, otherwise one could just say they're `@bugsbunny@warnerbros.com` in their bio and there's no way to verify that, plus there's gonna be duplicates problems (2 users claiming the same domain in their bio). Also yes, it's just awkward.

Of course. To use a custom _domain_ in a fediverse handle, [you have to serve the Webfinger redirects on it](https://fed.brid.gy/docs#fediverse-enhanced), at minimum, which you can only do if you own it.

@elfprince13 and I were talking about how to set a custom _username_ in a fediverse handle. I'll follow up a bit on that in #1190.

**Comment by frncs-eu** [comment]
> ^ done!
> 
> @frncs-eu:
> 
> > I don't think the bio approach is a good idea – ideally the user should verify the domain they're claiming, otherwise one could just say they're `@bugsbunny@warnerbros.com` in their bio and there's no way to verify that, plus there's gonna be duplicates problems (2 users claiming the same domain in their bio). Also yes, it's just awkward.
> 
> Of course. To use a custom _domain_ in a fediverse handle, [you have to serve the Webfinger redirects on it](https://fed.brid.gy/docs#fediverse-enhanced), at minimum, which you can only do if you own it.
> 
> @elfprince13 and I were talking about how to set a custom _username_ in a fediverse handle. I'll follow up a bit on that in #1190.

Sorry for the noise, all clear now! Thank you!

**Comment by snarfed** [comment]
Now that we support DMs/chat, we could let people DM us the domain they want to use for their bridged fediverse address. Easier than a custom authed UI.

**Comment by snarfed** [comment]
Realistically, Bluesky accounts bridged into the fediverse will always start on the @bsky.brid.gy domain. So, the next step here is either
1) Migrate the account from ...@bsky.brid.gy to @domain@domain with `Move`
OR
2) `Delete` the old account and start the new one on @domain@domain fresh.

1 is more work, but not too much, and obviously preferable.

**Comment by noamross** [comment]
I'm a little uncertain of the status of this feature.  Is it active? I see in docs that it says we can use our own domains, but it's unclear if the Bluesky => Fediverse route can be configured so one's fedi handle is `@domain.org@domain.org`.

**Comment by snarfed** [comment]
Hi Noam! No, we haven't built this yet. Thanks for the nudge on the docs, I should revise https://fed.brid.gy/docs#fediverse-enhanced to be clearer that fediverse custom domain handles are currently only supported for web sites, not Bluesky accounts. I'll do that.
